app/src/math/polygon2d.ts

```diff
@@ -0,0 +1,92 @@
+/**
+ * Merge nearby pixels if their distance is smaller than the threshold
+ *
+ * @param vertices
+ * @param threshold
+ */
+export function mergeNearbyVertices(
+  vertices: Array<[number, number]>,
+  threshold: number
+): Array<[number, number]> {
+  const newVertices: Array<[number, number]> = []
+  vertices.forEach((vertex) => {
+    if (newVertices.length === 0) {
+      newVertices.push(vertex)
+    } else {
+      const newVertex = newVertices[newVertices.length - 1]
+      const dist =
+        (vertex[0] - newVertex[0]) * (vertex[0] - newVertex[0]) +
+        (vertex[1] - newVertex[1]) * (vertex[1] - newVertex[1])
+      if (dist >= threshold * threshold) {
+        newVertices.push(vertex)
+      }
+    }
+  })
+  return newVertices
+}
+
+/**
+ * Determines if polygon has self-intersections and identifies all self-intersection coordinates
+ *
+ * @param vertices
+ */
+export function polyIsComplex(vertices: Array<[number, number]>): number[][] {
+  const intersections: number[][] = []
+
+  for (let i = 0; i < vertices.length - 1; i++) {
+    for (let j = i + 1; j < vertices.length - 1; j++) {
+      if (
+        intersects(
+          vertices[i],
+          vertices[i + 1],
+          vertices[j],
+          vertices[j + 1]
+        ) &&
+        j - i !== 1 &&
+        j - i !== vertices.length - 2
+      ) {
+        intersections.push([
+          vertices[i][0],
+          vertices[i][1],
+          vertices[i + 1][0],
+          vertices[i + 1][1],
+          vertices[j][0],
+          vertices[j][1],
+          vertices[j + 1][0],
+          vertices[j + 1][1]
+        ])
+      }
+    }
+  }
+
+  return intersections
+}
+
+/**
+ * Determines if two lines intersect
+ *
+ * @param v1
+ * @param v2
+ * @param v3
+ * @param v4
+ */
+function intersects(
```

Comment:
If v1=v3, v2=v4, you would treat this situation as a non-intersection.

Reply:
Fixed two checking issues:
- If `v1 == v3 && v2 == v4` then they will always intersect
- If lines are collinear (i.e. `det = 0 && v2 == v3` ) then they will also intersect, otherwise if they are any other parallel lines they will not intersect